### ISSUE TITLE
[Parity 11/16] Profiler and debug - chrome-trace, memory snapshot, record_function, NVTX/ITT, bottleneck

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#220]
Part of epic #209.

## Summary

Profiling is how users find performance problems. PyTorch's `torch.profiler` emits chrome-trace JSON + Kineto-compatible traces, reports per-op + per-kernel timings + memory, and integrates with TensorBoard. We have `PerformanceProfiler.cs`, `ProfilingCompiler.cs`, `CompilationBenchmark.cs` — partial coverage, no chrome-trace export guaranteed, no memory snapshot surface, no NVTX/ITT integration, no `record_function` API.

### `torch.profiler`
- `profiler.profile(activities=[CPU, CUDA], schedule=..., on_trace_ready=...)`
- `schedule(wait, warmup, active, repeat)` — skip, warmup, capture
- `record_function("name")` — user-annotated ranges in the trace
- Chrome-trace export (`.json.gz`) for `chrome://tracing` / Perfetto
- TensorBoard profiler plugin format
- Kineto integration: CUPTI events, NCCL events, CUDA kernel launches
- Stack trace per op (Python-side)

### Memory profiler
- `torch.cuda.memory._record_memory_history(enabled="all")` + `_dump_snapshot(path)`
- Memory timeline with allocation stacks (visualized in `memory_viz.py`)
- `memory_stats`, `memory_allocated`, `max_memory_allocated`, `memory_reserved`

### NVTX / ITT integration
- `torch.autograd.profiler.emit_nvtx()` — emit NVTX ranges for every autograd op
- Intel ITT integration for VTune

### Debugging aids
- `torch.autograd.detect_anomaly()` — backward NaN/Inf localization
- `torch.utils.bottleneck` — one-shot cProfile + autograd profiler + cudnn profiler

## Current state in the repo

- `Engines/Optimization/PerformanceProfiler.cs` — present; depth unclear.
- `Engines/Compilation/ProfilingCompiler.cs` — present.
- `Engines/Compilation/CompilationBenchmark.cs` — present.
- No chrome-trace export mentioned.
- No Kineto / CUPTI / NVTX wired through.
- No memory snapshot surface (see GPU primitives issue for CUDA pieces).
- Anomaly detection is in the autograd issue but also needs the profiler surface.

## How we beat PyTorch

1. **Profiler that's always-on at near-zero cost**: `TapeStepContext` already marks forward/backward boundaries. Our profiler samples via `TimestampContextSwitches` — no per-op `start_record/end_record` overhead like PyTorch. Target <1% tax when enabled.
2. **Chrome-trace + Perfetto-native export** plus a new **AiDotNet native trace** with richer info: per-pass optimization time, per-op dispatch tier (which tier of `BlasProvider` fired), autotune cache hit/miss. PyTorch can't surface this.
3. **Memory timeline that spans all allocators**: `TensorArena`, `GpuWorkspace`, `NativeMemoryManager`, per-backend caches. PyTorch's timeline covers only the CUDA caching allocator.
4. **NVTX + ITT auto-emit** (tied to GPU-primitives issue for NVTX) — zero user annotation.
5. **Record-function hierarchy via `using` scopes**: `using (Profiler.Range("myop"))` with zero overhead when disabled (compile-time check).
6. **Per-kernel attribution for fused ops**: when `PointwiseFusionPass` merges 5 ops, the trace still shows the 5 original logical names with a "fused" annotation. PyTorch shows only the fused kernel name.
7. **Chromium trace loadable in the same tab**: we emit `traceEvents` with the correct categories (`python_function`, `cpu_op`, `cuda_runtime`, `gpu_memcpy`, `gpu_memset`, `kernel`) so existing Perfetto UIs work unchanged.
8. **Bottleneck-style one-shot**: `dotnet ai-tensors bottleneck --script my_training.cs` runs a training iteration, captures chrome-trace + memory snapshot + autotune-miss list + SIMD dispatch stats. PyTorch's `torch.utils.bottleneck` is deprecated-feeling; ours is first-class.
9. **Anomaly localization that survives compile**: even with `LazyTensorScope` compiled plans, anomaly detection can localize a NaN to the forward source line via retained `TapeEntry` debug info.

### Profiler core
- [ ] `Profiler.Profile(options)` context with `Activities`, `Schedule`, `OnTraceReady`
- [ ] `Profiler.Schedule(wait, warmup, active, repeat)`
- [ ] `Profiler.Range(name)` / `Profiler.RecordFunction(name)` user annotations
- [ ] Per-op, per-fusion-group, per-pass timings
- [ ] Dispatch-tier attribution (which backend / which tier fired)
- [ ] Autotune hit/miss counters

### Trace export
- [ ] Chrome-trace `.json` / `.json.gz` export
- [ ] Perfetto-compatible export (same format + protocol buffer variant)
- [ ] TensorBoard profiler plugin format (uses Serialization issue's TensorBoard writer)
- [ ] AiDotNet native trace (optional richer format)

### Memory profiler
- [ ] `MemoryProfiler.RecordHistory(mode)` / `DumpSnapshot(path)`
- [ ] Memory timeline across all allocators
- [ ] Peak stats / reset / summary (ties to GPU-primitives issue)

### NVTX / ITT
- [ ] Auto-emit ranges from `LazyTensorScope` + `TapeStepContext` (implementation shared with GPU-primitives issue for the CUDA side)
- [ ] ITT integration for Intel VTune

### Debugging
- [ ] Anomaly mode (NaN/Inf localization with original forward call) — autograd issue scopes the trapping; this issue scopes the reporting
- [ ] `dotnet ai-tensors bottleneck` CLI

### Integration with compile
- [ ] Fused-op attribution in traces (original op names retained)
- [ ] Pass-by-pass timing for `Engines/Optimization/*` pipeline

## Acceptance criteria

- Profile a ResNet-50 training step, export chrome-trace, open in Perfetto, see per-op timings + cuDNN kernel launches + H2D/D2H copies.
- Memory snapshot during an OOM-adjacent training run shows the allocation timeline and largest live tensors, including `TensorArena`.
- NVTX ranges visible in Nsight Systems without user annotation.
- Record-function ranges nest correctly when nested.
- Overhead benchmark: enabled profiler adds ≤1% to a standard training step.
- `bottleneck` CLI produces a readable summary on a toy script.

## Implementation notes

- Build on existing `PerformanceProfiler.cs` + `ProfilingCompiler.cs`.
- Chrome-trace writer under `Engines/Profiling/Trace/ChromeTraceWriter.cs`.
- Memory timeline under `Engines/Profiling/Memory/`.
- Use `System.Diagnostics.ActivitySource` where natural so OpenTelemetry backends also work (bonus: lands on any dashboard that understands ActivitySource).

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state

| File | Status |
|---|---|
| `Engines/Optimization/PerformanceProfiler.cs` | Present; depth unclear |
| `Engines/Compilation/ProfilingCompiler.cs` | Present |
| `Engines/Compilation/CompilationBenchmark.cs` | Present |

**Missing:** chrome-trace/Perfetto export, Kineto/CUPTI integration, NVTX/ITT (except tied to #219), memory snapshot surface, `record_function` API, anomaly-mode reporting, `bottleneck` CLI.

## 2. Scope — full inventory

### Profiler core
```
Profiler.Profile(options) context
  - Activities: { CPU, CUDA, HIP, Metal, Vulkan, OpenCL, WebGPU }
  - Schedule(wait, warmup, active, repeat)
  - OnTraceReady callback
Profiler.Range(name) / Profiler.RecordFunction(name) user annotations
Per-op, per-fusion-group, per-pass timings
Dispatch-tier attribution (which BlasProvider tier fired)
Autotune hit/miss counters
```

### Trace export
```
ChromeTraceWriter (.json / .json.gz)
PerfettoWriter (protobuf variant)
TensorBoardProfilerPlugin format (uses #218's TensorBoard writer)
AiDotNetNativeTrace (optional richer format)
```

### Memory profiler
```
MemoryProfiler.RecordHistory(mode) / DumpSnapshot(path)
Memory timeline across TensorArena + GpuWorkspace + NativeMemoryManager + per-backend caches
Peak stats / reset / summary (ties to #219)
```

### NVTX / ITT
```
Auto-emit from LazyTensorScope + TapeStepContext (NVTX shared with #219)
ITT integration for Intel VTune
```

### Debugging
```
AnomalyMode reporting (NaN/Inf with forward-call-site localization)
(trapping is #214's, reporting is here)
```

### CLI
```
dotnet ai-tensors bottleneck --script my_training.cs
  → chrome-trace + memory snapshot + autotune-miss list + SIMD dispatch stats
```

### Integration with compile
```
Fused-op attribution in traces (original op names retained)
Pass-by-pass timing for Engines/Optimization/* pipeline
```

**Total: ~30 surfaces.**

## 3. Win vectors

1. **Always-on profiler at near-zero cost**: `TapeStepContext` marks boundaries; sample via `TimestampContextSwitches`. Target <1% tax.
2. **Chrome-trace + Perfetto-native + AiDotNet-native**: richer info than PyTorch — per-pass optimization time, per-op dispatch tier, autotune hit/miss.
3. **Memory timeline spans all allocators**: PyTorch covers CUDA caching allocator only.
4. **NVTX + ITT auto-emit** tied to #219 — zero annotation.
5. **Record-function hierarchy via `using` scopes** with compile-time zero-cost when disabled.
6. **Per-kernel attribution for fused ops**: trace shows all original logical names with "fused" annotation. PyTorch shows only fused kernel name.
7. **Chromium trace UI unchanged**: emit `traceEvents` with correct categories for existing Perfetto UIs.
8. **Bottleneck CLI first-class**: captures trace + memory snapshot + autotune-miss + SIMD dispatch stats.
9. **Anomaly localization survives compile**: even compiled plans retain `TapeEntry` debug info.

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| Profile ResNet-50 training step → chrome-trace → Perfetto | per-op timings + kernel launches + H2D/D2H visible |
| Memory snapshot during OOM-adjacent run | allocation timeline + largest live tensors incl. TensorArena |
| NVTX auto-emit in Nsight Systems | visible without user annotation |
| Record-function nested correctly | works |
| Profiler enabled overhead | ≤1% on training step |
| Bottleneck CLI | readable summary on toy script |
| Fused-op attribution | original names retained in trace |
| Pass timing | per-pass breakdown in trace |

## 5. File inventory

**New (~40):**
- `src/AiDotNet.Tensors/Engines/Profiling/` root
- `Profiling/Core/` — Profiler, Schedule, Activity, RecordFunction, DispatchTierAttribution (~10)
- `Profiling/Trace/` — ChromeTraceWriter, PerfettoWriter, TensorBoardPluginFormat, AiDotNetNativeTrace (~8)
- `Profiling/Memory/` — MemoryProfiler, MemoryTimeline, AllocSiteTracker (~6)
- `Profiling/Nvtx/` — NvtxAutoEmit (shared with #219) (~2)
- `Profiling/Itt/` — ItcIntegration for Intel VTune (~3)
- `Profiling/Anomaly/` — AnomalyReporter (forward-site localization) (~2)
- `tools/AiDotNet.Tensors.Cli/Bottleneck/` — CLI command (~5)

**Modified (~8):** `PerformanceProfiler.cs`, `ProfilingCompiler.cs`, `LazyTensorScope.cs`, `TapeStepContext.cs`, `CompiledInferencePlan.cs`, `CompiledTrainingPlan.cs`, `TensorArena.cs`, `GpuWorkspace.cs`.

**Tests (~20):** chrome-trace schema validation, Perfetto compatibility, memory-timeline correctness, NVTX visibility, record-function nesting, overhead measurement, bottleneck CLI output validation.

**Benchmarks (~3):** `ProfilerOverheadBenchmarks`, `MemoryTimelineBenchmarks`, `TraceExportBenchmarks`.

## 6. Acceptance gate

1. Profile ResNet-50 training step, export chrome-trace, open in Perfetto, see per-op + kernel + H2D/D2H.
2. Memory snapshot during OOM run lists allocation timeline + largest tensors including `TensorArena`.
3. NVTX ranges in Nsight Systems without user annotation.
4. Record-function ranges nest correctly.
5. Overhead ≤1% on standard training step.
6. `bottleneck` CLI produces readable summary on toy script.
7. Fused-op attribution preserves original names in traces.

## 7. Timeline — ~4 weeks

W1 Profiler core + Schedule + RecordFunction + DispatchTierAttribution.
W2 ChromeTraceWriter + PerfettoWriter + TensorBoardPlugin.
W3 MemoryProfiler + memory timeline across allocators.
W4 NVTX/ITT auto-emit + Bottleneck CLI + benchmarks + PR.

## 8. Dependencies

- **Blocks**: nothing hard.
- **Blocked by**: #219 (NVTX primitives), #218 (TensorBoard writer), #214 (AnomalyMode trapping).

## 9. Risks

- **Trace schema drift** — pin to stable `traceEvents` schema + Perfetto proto.
- **Memory timeline cost** when history recording enabled — opt-in.
- **ITT integration** requires Intel VTune SDK — document optional install.

---

*Part of epic #209. Plan ratified 2026-04-19.*
